fix(cli,ui): add consistent basePath handling for subpath deployments (N8N_PATH)

skyplane23 opened the pull request:
## Summary

This PR adds **first-class subpath support** by consistently honoring `N8N_PATH` (aka `basePath`) across the backend and frontend. It fixes the long-standing issue where n8n appears to work only with reverse-proxy rewrites (stripping the subpath) and breaks features that depend on exact URLs.

### What changed

**Backend (CLI/server):**

* Introduce and normalize `basePath` in `AbstractServer` and propagate it everywhere routes are registered.
  * Normalize (strip trailing `/`, ensure leading `/`) and reuse for all mounts.
* Serve static editor UI and assets under `basePath`:
  * `this.app.use('${basePath}/', historyApiHandler, express.static(...))`
  * Icon endpoints now resolve using `basePath` (`resolveIcon(basePath, packageName, req.originalUrl)`).
* REST/schema and internal routes now include `basePath`:
  * `/${basePath}/${restEndpoint}/schemas/:node/:version/:resource/:operation.json`
  * Health checks moved to `${basePath}/healthz` and `${basePath}/healthz/readiness`.
* Webhook-related mounts now include `basePath`:
  * Live forms: `${basePath}/${endpointForm}/*path`
  * Live webhooks: `${basePath}/${endpointWebhook}/*path`
  * Waiting forms/webhooks: `${basePath}/${endpointFormWaiting}`, `${basePath}/${endpointWebhookWaiting}`
  * MCP and MCP test: `${basePath}/${endpointMcp}/*path`, `${basePath}/${endpointMcpTest}/*path`
  * Test webhook DELETE: `${basePath}/${restEndpoint}/test-webhook/:id`
* Push/SSE/WebSocket paths include `basePath`:
  * Server-sent events / push: `${basePath}/${restEndpoint}/push`
  * WS upgrade check: `pathname.includes(\`/${restEndpoint}/push\`)`with`basePath\` awareness.
* Controller registry prefixes API routers with `basePath` (and normalizes it) instead of hardcoding root `/`.
* Auth service:
  * Add an additional skip-list check using `.some(skipEndpoint => endpoint.includes(skipEndpoint))` for robustness with `basePath`.
  * (OAuth callback list unchanged; still addressed by `restEndpoint`.)

**Frontend (editor-ui):**

* Fix authenticated redirect to avoid dropping the subpath:
  * Use `window.BASE_PATH` when building the fallback redirect.<br>`const redirect = to.query.redirect ?? encodeURIComponent(\`${url.pathname.replace(new RegExp(\`^${window.BASE_PATH}\`), '/')}${url.search}\`);\`

**Key effect:** All UI, API, push, schema, health, icons, and webhook endpoints are now mounted relative to `N8N_PATH`. Reverse-proxy rewrite hacks are no longer required for subpath deployments.

### Why

When n8n is hosted at `/my-n8n`, many routes ignored `N8N_PATH`. Users could mask this via proxy rewrites, but **Human-in-the-Loop (HITL)** and other URL-sensitive features would fail because the browser used a subpath URL while the server verified against a rewritten (root) URL. Centralizing `basePath` and applying it consistently prevents these mismatches.

---

## How to test

1. **Environment**

   ```env
   N8N_PATH=/my-n8n/
   N8N_EDITOR_BASE_URL=http://localhost:5678/my-n8n/
   WEBHOOK_URL=http://localhost:5678/my-n8n
   ```
2. **Start n8n** without any reverse-proxy path rewrites (no stripping `/my-n8n`).
3. **Editor UI**
   * Navigate to `http://localhost:5678/my-n8n/`
   * Confirm navigation/refresh and login redirects keep `/my-n8n`.
4. **Static assets & icons**
   * Check node/credential icons load from `${basePath}/icons/...`.
5. **Health endpoints**
   * `GET http://localhost:5678/my-n8n/healthz` → `{ status: "ok" }`
   * `GET http://localhost:5678/my-n8n/healthz/readiness` → 200 once DB ready.
6. **REST & schemas**
   * Schema JSON under `${basePath}/${restEndpoint}/schemas/...` resolves.
7. **Push/SSE/WS**
   * Open editor; network tab should show EventSource/WS to `${basePath}/${restEndpoint}/push`.
8. **Webhooks & forms (incl. HITL)**
   * Live/waiting form and webhook URLs include `${basePath}` and function.
   * **HITL**: Create a step requiring approval; open the approval link from the browser under `/my-n8n`. Signature verification should **succeed** (no mismatch).
9. **Test webhook deletion**
   * `DELETE ${basePath}/${restEndpoint}/test-webhook/:id` works.

---

## Breaking/behavior changes

* Health endpoints move under `${basePath}`:
  * `/healthz` → `${basePath}/healthz`
  * `/healthz/readiness` → `${basePath}/healthz/readiness`
* All editor assets, schemas, push, and webhook endpoints are now **basePath-prefixed**.<br>If you previously relied on root-mounted paths while also setting `N8N_PATH`, update any external monitors or allowlists accordingly.

**Backwards compatibility:**<br>With `N8N_PATH` unset or empty, normalized `basePath` is `''`, preserving current root behavior.

---

## Related Linear tickets, Github issues, and Community forum posts

`closes #19635` ([CAT-1439](https://linear.app/n8n/issue/CAT-1439/community-issue-subpath-deployment-via-n8n-path-breaks-endpoints)) — “Subpath deployment via N8N_PATH breaks: endpoints ignore basePath and Human-in-the-Loop fails”.

## Review / Merge checklist

- [X] PR title and summary are descriptive. ([conventions](<../blob/master/.github/pull_request_title_conventions.md>))
- [X] [Docs updated](<https://github.com/n8n-io/n8n-docs>) or follow-up ticket created.
- [X] Tests included.
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

skyplane23 committed:
feat(server): Introduce basePath configuration for endpoint routing

- Added basePath property to AbstractServer for dynamic endpoint management.
- Updated health check and webhook routes to utilize basePath.
- Modified ControllerRegistry to construct prefixes using basePath.
- Adjusted LoadNodesAndCredentials to resolve icons with basePath.
- Refactored Server class to pass basePath to various endpoint handlers.
- Enhanced Push class to support basePath in push endpoint setup.

skyplane23 committed:
feat(server): Introduce basePath configuration for endpoint routing

- Added basePath property to AbstractServer for dynamic endpoint management.
- Updated health check and webhook routes to utilize basePath.
- Modified ControllerRegistry to construct prefixes using basePath.
- Adjusted LoadNodesAndCredentials to resolve icons with basePath.
- Refactored Server class to pass basePath to various endpoint handlers.
- Enhanced Push class to support basePath in push endpoint setup.

skyplane23 committed:
feat(server): Introduce basePath configuration for endpoint routing

- Added basePath property to AbstractServer for dynamic endpoint management.
- Updated health check and webhook routes to utilize basePath.
- Modified ControllerRegistry to construct prefixes using basePath.
- Adjusted LoadNodesAndCredentials to resolve icons with basePath.
- Refactored Server class to pass basePath to various endpoint handlers.
- Enhanced Push class to support basePath in push endpoint setup.

skyplane23 committed:
fix(auth): Correct redirect URL handling in authenticated middleware to respect basePath

skyplane23 committed:
fix(auth): Correct redirect URL handling in authenticated middleware to respect basePath

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=19638) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=19638) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=19638) it.</sub>

skyplane23 committed:
refactor(push): Update setupPushServer to include basePath for websocket connection handling

skyplane23 committed:
refactor(push): Update setupPushServer to include basePath for websocket connection handling

skyplane23 committed:
refactor(auth): Update skipBrowserIdCheckEndpoints to include basePath for dynamic endpoint handling

skyplane23 committed:
refactor(auth): Update skipBrowserIdCheckEndpoints to include basePath for dynamic endpoint handling

skyplane23 committed:
refactor(auth): Update skipBrowserIdCheckEndpoints to include basePath for dynamic endpoint handling

skyplane23 committed:
refactor(auth): Update skipBrowserIdCheckEndpoints to include basePath for dynamic endpoint handling

skyplane23 committed:
fix(auth): Improve redirect URL handling in authenticated middleware to correctly account for basePath

skyplane23 committed:
fix(auth): Improve redirect URL handling in authenticated middleware to correctly account for basePath

skyplane23 committed:
fix(auth): Restore authentication middleware for protected type files in server

skyplane23 committed:
fix(auth): Restore authentication middleware for protected type files in server

skyplane23 committed:
fix(server): Adjust file serving to ensure correct path handling with basePath

skyplane23 committed:
fix(server): Adjust file serving to ensure correct path handling with basePath

skyplane23 committed:
fix(server): Enhance file serving logic to correctly handle relative paths with basePath

skyplane23 committed:
fix(server): Enhance file serving logic to correctly handle relative paths with basePath

skyplane23 committed:
fix(auth): Enhance redirect URL handling in authenticated middleware to correctly process basePath

skyplane23 committed:
fix(auth): Enhance redirect URL handling in authenticated middleware to correctly process basePath

skyplane23 committed:
fix(auth): Improve pathname normalization in authenticated middleware to handle basePath correctly

skyplane23 committed:
fix(auth): Improve pathname normalization in authenticated middleware to handle basePath correctly

skyplane23 committed:
fix(auth): Improve pathname normalization in authenticated middleware to handle basePath correctly

skyplane23 committed:
fix(auth): Improve pathname normalization in authenticated middleware to handle basePath correctly

skyplane23 committed:
fix(editor-ui): Update icon paths in NodeCreator and sanitizeHtml to use dynamic basePath

skyplane23 committed:
fix(editor-ui): Update icon paths in NodeCreator and sanitizeHtml to use dynamic basePath

skyplane23 committed:
fix(chat-trigger): Add instanceBaseUrl to createPage function for dynamic URL handling

skyplane23 committed:
fix(chat-trigger): Add instanceBaseUrl to createPage function for dynamic URL handling

tomi commented:
@skyplane23 we have decided the following:
1. Leave `N8N_PATH` as it is. No changes should be made in its behaviour.
2. Introduce a new `N8N_BASE_PATH` configuration option, which provides full support for hosting n8n under a subpath
3. `N8N_PATH` will be deprecated after `N8N_BASE_PATH` has been implemented

So to get this merged, the following changes are needed in your PR:
1. Get it in sync with latest changes, i.e. rebase on top of `master`
2. Introduce a new configuration option `N8N_BASE_PATH`. Make sure it gets normalized
4. Add a new service `PathResolvingService` that provides methods to resolve a full endpoint path, so it's encapsulated. Make sure it has tests
5. Make sure the base URL is reflected in `UrlService`. NOTE: `WEBHOOK_URL` should still override the whole webhook URL
6. Make sure auth cookies are scoped to this path

We also need e2e tests for this:
1. We need a separate config in our e2e tests that has this new env set, in `packages/testing/playwright/playwright-projects.ts`. Could be named e.g. `custom-base-path`
2. Create a new test suite under `packages/testing/playwright/tests/ui/` which has the `@mode:custom-base-path` in its name. Add some basic behaviour checks into this

Make sure the current behaviour with `N8N_PATH` isn't changed.

If you have any questions let me know!

skyplane23 committed:
feat(path-resolving): Introduce PathResolvingService for consistent endpoint path resolution across the application

- Added PathResolvingService to encapsulate logic for resolving full endpoint paths with the configured base path.
- Updated various components to utilize PathResolvingService for consistent URL construction, including server routes, public API endpoints, and static asset paths.
- Enhanced tests to verify correct behavior under custom base paths.

skyplane23 committed:
feat(path-resolving): Introduce PathResolvingService for consistent endpoint path resolution across the application

- Added PathResolvingService to encapsulate logic for resolving full endpoint paths with the configured base path.
- Updated various components to utilize PathResolvingService for consistent URL construction, including server routes, public API endpoints, and static asset paths.
- Enhanced tests to verify correct behavior under custom base paths.

skyplane23 committed:
feat(path-resolving): Introduce PathResolvingService for consistent endpoint path resolution across the application

- Added PathResolvingService to encapsulate logic for resolving full endpoint paths with the configured base path.
- Updated various components to utilize PathResolvingService for consistent URL construction, including server routes, public API endpoints, and static asset paths.
- Enhanced tests to verify correct behavior under custom base paths.

skyplane23 committed:
feat(path-resolving): Introduce PathResolvingService for consistent endpoint path resolution across the application

- Added PathResolvingService to encapsulate logic for resolving full endpoint paths with the configured base path.
- Updated various components to utilize PathResolvingService for consistent URL construction, including server routes, public API endpoints, and static asset paths.
- Enhanced tests to verify correct behavior under custom base paths.

skyplane23 committed:
fix(chat-trigger): Update createPage function and tests to use instanceBaseUrl for URL handling

- Set default value for instanceBaseUrl in createPage function to '/'.
- Refactor tests to replace basePath with instanceBaseUrl for clarity and consistency in redirect URL handling.

skyplane23 committed:
fix(chat-trigger): Update createPage function and tests to use instanceBaseUrl for URL handling

- Set default value for instanceBaseUrl in createPage function to '/'.
- Refactor tests to replace basePath with instanceBaseUrl for clarity and consistency in redirect URL handling.

tomi commented:
> Just following up on this. Any chance we can get this reviewed/merged when you have a moment?

Hi @skyplane23 

Thank you for making the changes and sorry for not getting back to you earlier. Yes, I'll give it a look this week. There seem to be couple conflicts. Would you mind resolving them 🙏

skyplane23 committed:
refactor(tests): Remove deprecated custom-base-path.spec.ts and add new isolated test for custom base path functionality

- Deleted the old custom-base-path.spec.ts file which contained outdated tests.
- Introduced a new test file that verifies n8n's behavior when deployed under a custom base path, ensuring UI and API functionality is intact.
- The new tests run in an isolated container to apply the custom base path environment variable.

skyplane23 committed:
fix(chat-trigger): Normalize instanceBaseUrl handling in createPage function

- Updated createPage function to ensure instanceBaseUrl is consistently normalized by adding a trailing slash if missing.
- Refactored related tests to verify correct URL handling for both absolute and relative instanceBaseUrl values, ensuring proper redirection and fetch requests.

skyplane23 committed:
Enhance base URL handling in createN8NStack to support custom base paths

- Updated the base URL construction to include a configurable base path, allowing n8n to be hosted under a custom path.
- Adjusted main container URL generation to append the base path, ensuring correct URL resolution for both UI and API interactions.

skyplane23 committed:
Enhance custom base path tests in Playwright

- Introduced new tests to verify n8n's behavior under a custom base path, including UI serving, health checks, and webhook trigger routes.
- Refactored existing tests to utilize a constant for the base path, improving maintainability and clarity.
- Ensured that routes are correctly bound to the configured base path, preventing accidental exposure at the root.

skyplane23 committed:
Refactor n8n wait strategy and enhance base path handling in Playwright tests

- Replaced the static wait strategy with a dynamic function to accommodate custom base paths in n8n.
- Updated navigation and API request handling to support base path configuration across various components.
- Improved test cases to validate behavior under custom base paths, ensuring correct URL resolution and functionality.

skyplane23 committed:
fix(LoadNodesAndCredentials): Normalize basePath handling for icon resolution

- Updated LoadNodesAndCredentials to handle root basePath ('/') correctly, preventing incorrect URL slicing.
- Added a test case to verify the correct behavior when resolving icons with a root basePath.
- Enhanced path resolution logic in PathResolvingService to accommodate customized health endpoints.

skyplane23 committed:
Enhance path normalization logic and add tests for trailing slashes

- Updated the normalizePathSegment function to strip multiple trailing slashes instead of just one, improving path normalization.
- Added new test cases to verify the correct behavior of normalizeBasePath when handling multiple trailing slashes.

tomi commented on packages/@n8n/config/src/index.ts:
```diff
@@ -132,6 +132,10 @@ export class GlobalConfig {
 	@Env('N8N_PATH')
 	path: string = '/';
 
+	/** Base path n8n is deployed to */
```
Can we also document whether one should configure this is `/n8n` or `n8n` 🙏

tomi commented on packages/@n8n/config/src/index.ts:
```diff
@@ -132,6 +132,10 @@ export class GlobalConfig {
 
+	/** Base path n8n is deployed to */
+	@Env('N8N_BASE_PATH')
+	basePath: string = '';
```
Let's also add a sanitize function that throws if both path and basePath are configured. That way we prevent users from configuring both. We should only allow one of these to be configured, with N8N_PATH being the deprecated one. Assume that if `path` is `/` it hasn't been configured

tomi commented on packages/@n8n/backend-common/src/utils/path-util.ts:
```diff
@@ -1,6 +1,39 @@
+ * normalizeBasePath('/prefix', '/app') // '/prefix/app'
+ * normalizeBasePath('prefix/', 'app/') // '/prefix/app'
+ */
+export function normalizeBasePath(basePath: string, urlPath: string): string {
```
After we have prevent configuring both `N8N_PATH` and `N8N_BASE_PATH` we shouldn't need this anymore

tomi commented on packages/@n8n/backend-common/src/__tests__/path-util.test.ts:
```diff
@@ -0,0 +1,106 @@
+		});
+	});
+
+	describe('isContainedWithin', () => {
```
We already have tests for these in `packages/@n8n/backend-common/src/utils/__tests__/path-util.test.ts`. No need to add new ones

tomi commented on packages/@n8n/backend-common/src/__tests__/path-util.test.ts:
```diff
@@ -0,0 +1,106 @@
+		});
+	});
+
+	describe('safeJoinPath', () => {
```
We already have tests for these in `packages/@n8n/backend-common/src/utils/__tests__/path-util.test.ts`. No need to add new ones

tomi commented on packages/cli/src/deprecation/deprecation.service.ts:
```diff
@@ -73,6 +73,12 @@ export class DeprecationService {
 			checkValue: (value: string | undefined): value is 'own' => value === 'own',
 		},
+		{
+			envVar: 'N8N_PATH',
```
👍

tomi commented on packages/cli/src/services/path-resolving.service.ts:
```diff
@@ -0,0 +1,258 @@
+ * Service for resolving full endpoint paths with the configured base path.
+ * Encapsulates path resolution logic for consistent URL construction across the application.
+ *
+ * This is the single source of truth for all path resolution in n8n.
```
👏

tomi commented on packages/cli/src/services/path-resolving.service.ts:
```diff
@@ -0,0 +1,258 @@
+	 */
+	resolveEndpoint(endpoint: string): string {
+		// Remove leading slash from endpoint if present
+		const normalizedEndpoint = endpoint.startsWith('/') ? endpoint.slice(1) : endpoint;
```
This is repeated in many places. Can we extract this to `stripTrailingSlash`

tomi commented on packages/cli/src/services/path-resolving.service.ts:
```diff
@@ -0,0 +1,258 @@
+		if (!normalizedPath) {
+			return this.resolveEndpoint(restEndpoint);
+		}
+		return this.resolveEndpoint(`${restEndpoint}/${normalizedPath}`);
```
This repeats in all these methods here. Could we extract it 🙏

tomi commented on packages/cli/src/push/index.ts:
```diff
@@ -73,11 +74,13 @@ export class Push extends TypedEmitter<PushEvents> {
 
 	/** Sets up the main express app to upgrade websocket connections */
-	setupPushServer(restEndpoint: string, server: Server, app: Application) {
+	setupPushServer(pathResolvingService: PathResolvingService, server: Server, app: Application) {
```
Could we pass only the pushEndpoint here instead the entire PathResolvingService

tomi commented on packages/cli/src/push/index.ts:
```diff
@@ -97,9 +100,9 @@ export class Push extends TypedEmitter<PushEvents> {
 
 	/** Sets up the push endpoint that the frontend connects to. */
-	setupPushHandler(restEndpoint: string, app: Application) {
+	setupPushHandler(pathResolvingService: PathResolvingService, app: Application) {
```
Same here: Could we pass only the pushEndpoint here instead the entire PathResolvingService?

tomi reviewed (commented):
Thank you so much for the PR @skyplane23 !

I went through it and it works pretty well 👏 There are couple comments regarding the implementation.

I also found couple things when testing it:
1. The FE periodically checks the health endpoint, and that doesn't seem to work, which cause n8n to show 

<img width="175" height="93" alt="Image" src="https://github.com/user-attachments/assets/2d4f3828-faa3-4ba4-950b-da180d8fafeb" />

in the workflow editor. This is likely fixed by changing the code in `packages/cli/src/utils/health-endpoint.util.ts:24`

2. The Public API doesn't seem to work. Something wrong in the routing?

I tested it with env vars
```
export N8N_BASE_PATH=n8n
export WEBHOOK_URL="http://localhost:8080/n8n"
```

and having an nginx on port 8080 route all requests to /n8n to the instance

<details>
<summary>nginx.conf</summary>
```
map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      close;
}

server {
    listen 80;
    server_name localhost;

    location = / {
        return 302 /n8n/;
    }

    location /n8n/ {
        proxy_pass http://host.docker.internal:5678;

        proxy_http_version 1.1;
        proxy_set_header Host              $host;
        proxy_set_header X-Real-IP         $remote_addr;
        proxy_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-Host  $host;

        proxy_set_header Upgrade    $http_upgrade;
        proxy_set_header Connection $connection_upgrade;

        proxy_buffering    off;
        proxy_cache        off;
        proxy_read_timeout 3600s;
        proxy_send_timeout 3600s;

        client_max_body_size 100m;
    }
}
```
</details>

shortstacked reviewed (commented):
Safe to approve from e2e/testing side.

skyplane23 committed:
Refactor controller registry and update icon resolution test

- Removed unused Logger import from controller.registry.ts to streamline the code.
- Updated the icon resolution test to correctly handle the root basePath, ensuring accurate URL generation for icons.

skyplane23 committed:
feat(path-util): Enhance path normalization and introduce new assertions

- Added `assertPathAndBasePathAreNotBothSet` to enforce configuration rules for `N8N_PATH` and `N8N_BASE_PATH`.
- Updated `normalizeBasePath` to improve handling of base paths, ensuring correct normalization.
- Removed outdated tests and added new tests for `normalizeBasePath` and the new assertion function, verifying behavior under various configurations.
- Refactored related components to utilize the updated path resolution logic, ensuring consistent URL handling across the application.

skyplane23 committed:
feat(config): Add basePath to GlobalConfig and update related tests

- Introduced basePath property to GlobalConfig across multiple test files to ensure consistent handling of base paths.
- Updated mock configurations in tests to include basePath, enhancing test coverage for path-related functionality.
- Ensured that all relevant services and commands reflect the new basePath property for improved URL resolution.

skyplane23 committed:
fix(server): Update endpoint middleware to resolve paths correctly under custom base paths

- Modified the server to mount the overwrite endpoint middleware using the resolved path, ensuring compatibility when n8n is hosted under a custom base path.
- Removed hardcoded path usage to enhance flexibility and maintainability in URL handling.

skyplane23 committed:
fix(auth): Update authorization endpoint to use dynamic prefix for path resolution

- Modified the authorization endpoint in AuthService to utilize a dynamic prefix, ensuring correct URL generation under custom base paths.
- This change enhances flexibility and maintains consistency with recent updates to base path handling across the application.

skyplane23 committed:
fix(controller): Refactor return URL generation to utilize dynamic base path resolution

- Updated the DynamicCredentialsController to use PathResolvingService for constructing return URLs, ensuring correct handling of base paths.
- Enhanced tests to verify the behavior of the sign-in return URL generation, preventing duplication of the base path in the final URL.

skyplane23 committed:
fix(server): Update webhook endpoint to use dynamic path resolution

- Modified the AbstractServer class to utilize PathResolvingService for generating the webhook waiting endpoint, enhancing flexibility in URL handling under custom base paths.

skyplane23 committed:
feat(markdown): Implement dynamic base path handling for static images

- Added support for rendering static image URLs based on the configured base path in the N8nMarkdown component.
- Introduced a new global type declaration for BASE_PATH to facilitate dynamic path resolution.
- Enhanced tests to verify correct rendering of images under both root and configured base paths.

skyplane23 committed:
fix(auth): Resolve merge conflicts in AuthService for browserId check logic

- Cleaned up the conditional logic for skipping the browserId check in the AuthService.
- Ensured consistent handling of endpoints that bypass the browserId check, improving code clarity and maintainability.
